sympy/combinatorics/free_group.py

```diff
@@ -0,0 +1,938 @@
+        """This is called the External Representation of `FreeGroupElm`
+        """
+        return tuple(flatten(self.array_form))
+
+    def __str__(self):
+        if self.is_identity:
+            return "<identity>"
+
+        symbols = self.group.symbols
+        str_form = ""
+        array_form = self.array_form
+        for i in range(len(array_form)):
+            if i == len(array_form) - 1:
+                if array_form[i][1] == 1:
+                    str_form += str(array_form[i][0])
+                else:
+                    str_form += str(array_form[i][0]) + \
+                                    "**" + str(array_form[i][1])
+            else:
+                if array_form[i][1] == 1:
+                    str_form += str(array_form[i][0]) + "*"
+                else:
+                    str_form += str(array_form[i][0]) + \
+                                    "**" + str(array_form[i][1]) + "*"
+        return str_form
+
+    __repr__ = __str__
+
+    def __pow__(self, n):
+        n = as_int(n)
+        group = self.group
+        if n == 0:
+            return group.identity
+
+        if n < 0:
+            n = -n
+            return (self.inverse())**n
+
+        result = self
+        for i in range(n - 1):
+            result = result*self
+        # this method can be improved instead of just returning the
+        # multiplication of elements
+        return result
+
+    def __mul__(self, other):
+        """Returns the product of elements belonging to the same `FreeGroup`.
+
+        Examples
+        ========
+
+        >>> from sympy.combinatorics.free_group import free_group
+        >>> f, x, y, z = free_group("x y z")
+        >>> x*y**2*y**-4
+        x*y**-2
+        >>> z*y**-2
+        z*y**-2
+        >>> x**2*y*y**-1*x**-2
+        <identity>
+
+        """
+        group = self.group
+        if not isinstance(other, group.dtype):
+            raise TypeError("only FreeGroup elements of same FreeGroup can "
+                    "be multiplied")
+        if self.is_identity:
+            return other
+        if other.is_identity:
+            return self
+        r = list(self.array_form + other.array_form)
+        zero_mul_simp(r, len(self.array_form) - 1)
+        return group.dtype(tuple(r))
+
+    def __div__(self, other):
+        group = self.group
+        if not isinstance(other, group.dtype):
+            raise TypeError("only FreeGroup elements of same FreeGroup can "
+                    "be multiplied")
+        return self*(other.inverse())
+
+    def __rdiv__(self, other):
+        group = self.group
+        if not isinstance(other, group.dtype):
+            raise TypeError("only FreeGroup elements of same FreeGroup can "
+                    "be multiplied")
+        return other*(self.inverse())
+
+    __truediv__ = __div__
+
+    __rtruediv__ = __rdiv__
+
+    def __add__(self, other):
+        raise TypeError("unsupported operand type(s) for + or add: '%s' and '%s'"
```

Comment:
The correct behavior here (and possibly in other cases where you're raising `TypeError`) is to `return NotImplemented`. That way, other classes can define `__add__` and `__radd__` to add FreeGroupElm objects.